From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xiaoxia Lang <xxlang@grandstream.cn>
Date: Fri, 18 Aug 2023 15:06:27 +0800
Subject: fix: madvise fail causing crash.

Call munlock if madvise failed.

diff --git a/base/allocator/partition_allocator/page_allocator_internals_posix.h b/base/allocator/partition_allocator/page_allocator_internals_posix.h
index fb3be9054da3b5e4551ce3d7922e27208766a5af..7e4d47c8c5a19a5a2f136d7f18bd0941efb44728 100644
--- a/base/allocator/partition_allocator/page_allocator_internals_posix.h
+++ b/base/allocator/partition_allocator/page_allocator_internals_posix.h
@@ -361,7 +361,13 @@ void DiscardSystemPagesInternal(uintptr_t address, size_t length) {
   // performance benefits unclear.
   //
   // Therefore, we just do the simple thing: MADV_DONTNEED.
-  PA_PCHECK(0 == madvise(ptr, length, MADV_DONTNEED));
+  int ret = madvise(ptr, length, MADV_DONTNEED);
+  if (ret) {
+    // MADV_DONTNEED will fail if the specified address range includes locked pages.
+    munlock(ptr, length);
+    ret = madvise(ptr, length, MADV_DONTNEED);
+  }
+  PA_PCHECK(ret == 0);
 #endif
 }
 
